### PR TITLE
Revamping the modules stack to support explicit frame passing actions

### PR DESCRIPTION
**Pull request by aseemr**
In support of `witness_h_exists`. Will be doing some cleanup (comments etc.) in the next day or two.

In the heap module, there are two types of actions: one that quantify over the frame (the existing ones), and one that take frame as an explicit argument.

In `Mem` the actions are only of the latter type (in the interface). And then the semantics and the modules up the stack are also modified as such.

### CODE REVIEW

**Comment by mtzguido** [comment]
Seems fine to me after a quick look. Not sure if you mentioned this in the meeting last Wednesday, but why retain the old `action` too alongside `action_with_frame`? Is it just to not explicitly refactor them right now?

Also, what do you think about wrapping the frame with `erased`? Currently it's about the same since it's a `... -> prop`, but it would make explicit that it's computationally irrelevant.

**Comment by aseemr** [comment]
Thanks for the comments Guido!

The two types of actions are only in the `Heap` module, `Memory` and above, the actions are frame passing only. The advantage is as you point out: I did not have to touch most of the existing actions in the heap module, rather I could define a lift from quantified frame actions to explicit frame actions in Memory and that's all. The only action that required changes in the heap module is `witness_h_exists`.

I actually like the `frame` as is. As you say, this is already an `slprop`, so perhaps clear that it is computationally irrelevant. Also, the frame does not show up beyond the effect. For existing APIs too we have been using non-erased frame arguments. Happy to discuss more.

File: ulib/experimental/Steel.Semantics.Hoare.MST.fst [hunk 1]
```diff
@@ -773,123 +786,123 @@ let depends_only_on2_commutes_with_weaker
...
+//       m0;
+//     commute_assoc_star_right (st.locks_invariant m1) frame post frame';
+//     apply_interp_ext
+//       (st.locks_invariant m1 `st.star` (post `st.star` (frame `st.star` frame')))
+//       (st.locks_invariant m1 `st.star` ((frame `st.star` post) `st.star` frame'))
+//       m1;
+//     weaken_fp_prop frame frame' m0 m1
+//   in
+//   ()
+// #pop-options
```
Comment (nikswamy) [inline ulib/experimental/Steel.Semantics.Hoare.MST.fst:905]: Any reason to retain this large block of commented proof?

File: ulib/experimental/Steel.Semantics.Hoare.MST.fst [hunk 2]
```diff
@@ -1048,19 +1062,29 @@ let par_weaker_lpre_and_stronger_lpost_r
...
-  commute_star_right (st.locks_invariant state) preL preR;
-  apply_interp_ext
-    (st.locks_invariant state `st.star` (preL `st.star` preR))
-    (st.locks_invariant state `st.star` (preR `st.star` preL))
-    state;
+  // commute_star_right (st.locks_invariant state) preL preR;
+  // apply_interp_ext
+  //   (st.locks_invariant state `st.star` (preL `st.star` preR))
+  //   (st.locks_invariant state `st.star` (preR `st.star` preL))
+  //   state;
```
Comment (nikswamy) [inline ulib/experimental/Steel.Semantics.Hoare.MST.fst:1069]: can it go?

File: ulib/experimental/Steel.Semantics.Hoare.MST.fst [hunk 3]
```diff
@@ -1424,24 +1478,18 @@ let rec run
...
-    let m0 = get () in
-    let Step new_pre new_post _ _ f = step f in
-    let m1 = get () in
-    let x = run f in
-    let m2 = get () in
-    preserves_frame_trans pre new_pre (new_post x) m0 m1 m2;
-    preserves_frame_stronger_post pre post new_post x m0 m2;
-    x
+    let Step _ _ _ _ f = step frame f in
+    run frame f
```
Comment (nikswamy) [inline ulib/experimental/Steel.Semantics.Hoare.MST.fst:1495]: nice simplification

File: ulib/experimental/Steel.Memory.fsti [hunk 4]
```diff
@@ -498,3 +497,17 @@ val star_is_witinv_right (#a:Type)
   : Lemma (requires (is_witness_invariant g))
           (ensures  (is_witness_invariant (fun x -> f x `star` g x)))
           //[SMTPat   (is_witness_invariant (fun x -> f x `star` g x))]
+
+let equiv_ext_right (p q r:slprop)
+  : Lemma
+      (requires q `equiv` r)
+      (ensures (p `star` q) `equiv` (p `star` r))
```
Comment (nikswamy) [inline ulib/experimental/Steel.Memory.fsti:504]: Surprising to have this lemma here ... it's quite specific.

File: ulib/experimental/Steel.Memory.fst [hunk 5]
```diff
@@ -323,14 +323,48 @@ let mem_evolves =
...
 ////////////////////////////////////////////////////////////////////////////////
 // Lifting heap actions
 ////////////////////////////////////////////////////////////////////////////////
 
-let tot_pre_action_except (e:inames) (fp:slprop u#a) (a:Type u#b) (fp':a -> slprop u#a) =
-  hmem_with_inv_except e fp -> (x:a & hmem_with_inv_except e (fp' x))
+(**
+  Heap actions come in two equivalent flavors:
+  - one that take an explicit frame as argument (action_with_frame)
+  - and one that quantify over all the grame
```
Comment (nikswamy) [inline ulib/experimental/Steel.Memory.fst:356]: typo in comment

File: ulib/experimental/Steel.Memory.fst [hunk 6]
```diff
@@ -323,14 +323,48 @@ let mem_evolves =
...
+  NMSTTotal.NMSTATETOT a (full_mem u#1) mem_evolves
+    (requires fun m0 ->
+        inames_ok except m0 /\
+        interp (expects `star` locks_invariant except m0) m0)
+    (ensures fun m0 x m1 ->
+        inames_ok except m1 /\
+        interp (provides x `star` locks_invariant except m1) m1 /\
+        preserves_frame except expects (provides x) m0 m1)
+
+let action_except_nf (a:Type u#a) (except:inames) (expects:slprop) (provides: a -> slprop) =
```
Comment (nikswamy) [inline ulib/experimental/Steel.Memory.fst:346]: this definition doesn't seem to be used?

File: ulib/experimental/Steel.Memory.fsti [hunk 7]
```diff
@@ -434,10 +433,10 @@ let is_frame_monotonic #a (p : a -> slprop) : prop =
 let is_witness_invariant #a (p : a -> slprop) =
   forall x y m. interp (p x) m /\ interp (p y) m ==> x == y
 
-val witness_h_exists (#opened_invariants:_) (#a:_) (p:(a -> slprop){is_frame_monotonic p})
+val witness_h_exists (#opened_invariants:_) (#a:_) (p:a -> slprop)
```
Comment (nikswamy) [inline ulib/experimental/Steel.Memory.fsti:436]: nice, the payoff!

File: ulib/experimental/Steel.Heap.fsti [hunk 8]
```diff
@@ -578,8 +604,15 @@ val witinv_framon (#a:_) (p : a -> slprop)
...
-  : action (h_exists p) (erased a) (fun x -> p x)
+
+(**
+  witness_h_exists is defined with action_with_frame as it allows us to define it with any p
+
+  With the quantified frame actions, it creates an issue, since we have to prove that the witness is ok for all frames
+    whereas with an explicit frame, we can pick the witness for that particular frame
+*)
+val witness_h_exists (#a:_) (p:a -> slprop)
+  : action_with_frame (h_exists p) (erased a) (fun x -> p x)
```
Comment (nikswamy) [inline ulib/experimental/Steel.Heap.fsti:615]: nice

File: ulib/experimental/Steel.Effect.fst [hunk 9]
```diff
@@ -109,25 +109,25 @@ let repr a pre post req ens =
 let returnc (a:Type u#a) (x:a) (p:a -> slprop) = fun _ -> x
 
 let bind a b pre_f post_f req_f ens_f post_g req_g ens_g f g
-  = fun m0 ->
-      let x = f () in
-      g x ()
+  = fun frame ->
+      let x = f frame in
+      g x frame
```
Comment (nikswamy) [inline ulib/experimental/Steel.Effect.fst:114]: Surprising diff ... I guess the original `m0` was poorly named, it's just a unit.

**Comment by nikswamy** [comment]
Thanks Aseem. This looks really good, overall. Nice to have the explicit frames and to see witness_h_exists working with a weaker precondition.

In the long run, it might be helpful for us to just settle on frame passing actions everywhere, rather than defining them initially as preserving all frames non-constructively and then lifting. It might even make some of those initial frame preservation proof a bit easier and easier to maintain. But, that's something we can attempt later.

Also, since slprop is already marked `erasable`, I think we shouldn't need to explicitly have an `frame:erased slprop`.

**Pull request merged**